alf/networks/normalizing_flow_networks.py

```diff
@@ -0,0 +1,408 @@
+
+
+class NormalizingFlowNetwork(Network):
+    """The base class for normalizing flow networks.
+
+    Compared to traditional ``Network`` classes, its subclass needs to implement
+    the interface ``make_invertible_transform()``.
+    """
+
+    def __init__(
+            self,
+            input_tensor_spec: alf.nest.NestedTensorSpec,
+            conditional_input_tensor_spec: alf.nest.NestedTensorSpec = None,
+            use_transform_cache: bool = True,
+            name: str = "NormalizingFlowNetwork"):
+        """
+        Args:
+            input_tensor_spec: a rank-1 tensor spec
+            conditional_input_tensor_spec: a nested tensor spec
+            use_transform_cache: whether use cached transform. When there
+                is a conditional input, different transforms might be created
+                depending on the conditonal inputs. When there is no conditional
+                input, the same transform will always be used.
+                Note that this only caches the transform itself; to correctly
+                cache the inverse result, you also have to set ``cache_size=1``
+                when creating the transform.
+            name: name of the network
+        """
+        if conditional_input_tensor_spec is None:
+            super().__init__(input_tensor_spec, name=name)
+            self._conditional_inputs = False
+        else:
+            super().__init__(
+                (input_tensor_spec, conditional_input_tensor_spec), name=name)
+            self._conditional_inputs = True
+        self._use_transform_cache = use_transform_cache
+        self._cached_transform = (None, None)
+
+    @property
+    def use_conditional_inputs(self):
+        """
+        Returns:
+            bool: Whether this normalizing flow uses inputs to condition the
+                transforms.
+        """
+        return self._conditional_inputs
+
+    def make_invertible_transform(
+            self, conditional_inputs: alf.nest.NestedTensor = None):
+        r"""Express the network forward computation as an invertible PyTorch
+        ``Transform``. This overall transformation can be a composed one chaining
+        many transformation layers.
+
+        Args:
+            conditional_inputs: an optional nested conditional inputs that
+                condition the mapping :math:`x \rightarrow y`.
+
+        Returns:
+            torch.td.Transform: an invertible transform
+        """
+        if not self._use_transform_cache:
+            return self._make_invertible_transform(conditional_inputs)
+        old_t, old_z = self._cached_transform
+        if conditional_inputs is old_z and old_t is not None:
+            return old_t
+        t = self._make_invertible_transform(conditional_inputs)
+        self._cached_transform = (t, conditional_inputs)
+        return t
+
+    def _make_invertible_transform(
+            self, conditional_inputs: alf.nest.NestedTensor = None):
```

Comment:
This function should be dcumented since subclass needs to implement this function.